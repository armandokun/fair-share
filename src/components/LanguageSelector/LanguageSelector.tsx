import {
  FormControl,
  MenuItem,
  Select,
  SelectChangeEvent,
} from "@mui/material";
import { useTranslation } from "react-i18next";

const LanguageSelector = () => {
  const { i18n } = useTranslation();

  const selectedLanguage = i18n.language;
  const availableLanguages = Object.keys(i18n.services.resourceStore.data);

  const handleChange = (event: SelectChangeEvent) => {
    i18n.changeLanguage(event.target.value);
  };

  const getLanguageName = (language: string) =>
    language === "en" ? "EN" : "LT";

  const renderLanguageItems = () =>
    availableLanguages.map((language) => (
      <MenuItem value={language} key={language}>
        {getLanguageName(language)}
      </MenuItem>
    ));

  return (
    <FormControl>
      <Select
        id="language-select"
        value={selectedLanguage}
        onChange={handleChange}
      >
        {renderLanguageItems()}
      </Select>
    </FormControl>
  );
};

export default LanguageSelector;
