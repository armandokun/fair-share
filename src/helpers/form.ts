export const getSalaryPercent = (
  salaries: Array<number>,
  rentToPay: number
) => {
  const totalSalary = salaries.reduce(
    (accumulator, currentNumber) => accumulator + currentNumber,
    0
  );

  return (rentToPay / totalSalary) * 100;
};

export const getIndividualPayment = (
  salary: number,
  salaryPercent: number
) => {
  const individualPayment = (salaryPercent / 100) * salary

  return individualPayment.toFixed(2)
};

export const decimalCount = (number: number) => {
  const numberAsString = number.toString();

  if (numberAsString.includes('.')) {
    return numberAsString.split('.')[1].length;
  }

  return 0;
}
