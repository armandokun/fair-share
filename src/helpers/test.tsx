import { I18nextProvider } from "react-i18next";

import i18n from "../translations";

export const withIntl = (children: JSX.Element) => {
  return <I18nextProvider i18n={i18n}>{children}</I18nextProvider>;
};
