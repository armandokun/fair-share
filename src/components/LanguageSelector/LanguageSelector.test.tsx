import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import LanguageSelector from ".";

import { withIntl } from "../../helpers/test";
import i18n from "../../translations";

describe("<LanguageSelector />", () => {
  const renderHelper = (element: JSX.Element) => render(withIntl(element));

  const getItem = jest.spyOn(window.localStorage, "getItem");
  const setItem = jest.spyOn(window.localStorage, "setItem");

  beforeEach(() => {
    getItem.mockImplementation(() => null);
    setItem.mockImplementation(() => null);
  });

  it("renders correctly", () => {
    const { container } = renderHelper(<LanguageSelector />);

    expect(container).toMatchSnapshot();
  });

  it("calls local storage to check for saved language", () => {
    renderHelper(<LanguageSelector />);

    expect(getItem).toHaveBeenCalledTimes(1);
    expect(getItem).toHaveBeenCalledWith("language");
  });

  it("renders with default value", () => {
    renderHelper(<LanguageSelector />);

    expect(screen.getByText("EN")).toBeInTheDocument();
  });

  it("renders different selected language when local storage finds saved value", () => {
    getItem.mockImplementation(() => "lt");

    renderHelper(<LanguageSelector />);

    expect(screen.getByText("LT")).toBeInTheDocument();
  });

  it("renders language items when clicked on select component", () => {
    const { container } = renderHelper(<LanguageSelector />);

    userEvent.click(screen.getByText("EN"));

    expect(container).toMatchSnapshot();
  });

  it("sets different language when selected", () => {
    renderHelper(<LanguageSelector />);

    userEvent.click(screen.getByText("EN"));

    userEvent.click(screen.getByText("LT"));

    expect(setItem).toHaveBeenCalledTimes(2);
    expect(setItem).toHaveBeenCalledWith("language", "lt");
  });

  it("sets same language when selected", () => {
    renderHelper(<LanguageSelector />);

    userEvent.click(screen.getByTestId("language-select"));

    userEvent.click(screen.getByText("EN"));

    expect(setItem).toHaveBeenCalledTimes(0);
  });
});
