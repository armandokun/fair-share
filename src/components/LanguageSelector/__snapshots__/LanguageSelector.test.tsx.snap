// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageSelector /> renders correctly 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <div
      aria-label="language-select"
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      data-testid="language-select"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="language-select"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="language-select"
        role="button"
        tabindex="0"
      >
        EN
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value="en"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-hdw1oc"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`<LanguageSelector /> renders language items when clicked on select component 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <div
      aria-label="language-select"
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      data-testid="language-select"
    >
      <div
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-labelledby="language-select"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="language-select"
        role="button"
        tabindex="0"
      >
        EN
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value="en"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined MuiSelect-iconOpen css-bpeome-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-hdw1oc"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
